OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -0,0 +1,1446 @@
+      .WillOnce(Invoke(mock_intern_string_if_necessary_and_get_key));
+  EXPECT_CALL(*producer_, EnqueueCaptureEvent)
+      .Times(2)
+      .WillRepeatedly(Invoke(mock_enqueue_capture_event));
+
+  Color expected_color{1.f, 0.8f, 0.6f, 0.4f};
+
+  pid_t tid = GetCurrentThreadId();
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkDebugMarkerBegin(command_buffer_, text, expected_color);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+  uint64_t pre_submit_time_1 = MonotonicTimestampNs();
+  std::optional<QueueSubmission> queue_submission_optional_1 =
+      tracker_.PersistCommandBuffersOnSubmit(1, &submit_info_);
+  tracker_.PersistDebugMarkersOnSubmit(queue_, 1, &submit_info_, queue_submission_optional_1);
+  uint64_t post_submit_time_1 = MonotonicTimestampNs();
+  tracker_.CompleteSubmits(device_);
+  tracker_.ResetCommandBuffer(command_buffer_);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkDebugMarkerEnd(command_buffer_);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+  std::optional<QueueSubmission> queue_submission_optional_2 =
+      tracker_.PersistCommandBuffersOnSubmit(1, &submit_info_);
+  tracker_.PersistDebugMarkersOnSubmit(queue_, 1, &submit_info_, queue_submission_optional_2);
+  tracker_.CompleteSubmits(device_);
+
+  EXPECT_THAT(actual_reset_slots_1, UnorderedElementsAre(kSlotIndex1, kSlotIndex3));
+  EXPECT_THAT(actual_reset_slots_2,
+              UnorderedElementsAre(kSlotIndex2, kSlotIndex4, kSlotIndex5, kSlotIndex6));
+  ASSERT_EQ(actual_capture_events.size(), 2);
+
+  const orbit_grpc_protos::CaptureEvent& actual_capture_event_1 = actual_capture_events.at(0);
+  EXPECT_TRUE(actual_capture_event_1.has_gpu_queue_submission());
+  const orbit_grpc_protos::GpuQueueSubmission& actual_queue_submission_1 =
+      actual_capture_event_1.gpu_queue_submission();
+  EXPECT_EQ(actual_queue_submission_1.num_begin_markers(), 1);
+  EXPECT_EQ(actual_queue_submission_1.completed_markers_size(), 0);
+
+  const orbit_grpc_protos::CaptureEvent& actual_capture_event_2 = actual_capture_events.at(1);
+  EXPECT_TRUE(actual_capture_event_2.has_gpu_queue_submission());
+  const orbit_grpc_protos::GpuQueueSubmission& actual_queue_submission_2 =
+      actual_capture_event_2.gpu_queue_submission();
+  EXPECT_EQ(actual_queue_submission_2.num_begin_markers(), 0);
+  EXPECT_EQ(actual_queue_submission_2.completed_markers_size(), 1);
+  const orbit_grpc_protos::GpuDebugMarker& actual_debug_marker =
+      actual_queue_submission_2.completed_markers(0);
+
+  ExpectDebugMarkerEndEq(actual_debug_marker, kTimestamp5, expected_text_key, expected_color, 0);
+  ExpectDebugMarkerBeginEq(actual_debug_marker, kTimestamp2, pre_submit_time_1, post_submit_time_1,
+                           tid);
+}
+
+TEST_F(SubmissionTrackerTest, CanRetrieveDebugMarkerAcrossTwoSubmissionsEvenWhenNotCapturingBegin) {
```

Comment:
Maybe "when begin not captured" is slightly better?

Reply:
Done.